merlin-server/sql/merlin/tables/span.sql

```diff
@@ -42,3 +32,39 @@ comment on column span.type is e''
   'The type of span, implying the shape of its attributes.';
 comment on column span.attributes is e''
   'A set of named values annotating this span as a whole.';
+
+create or replace function span_integrity_function()
+  returns trigger
+  security invoker
+  language plpgsql as $$begin
+  if not exists(select from dataset where dataset.id = new.dataset_id for key share of dataset)
+  then
+    raise exception 'foreign key violation: there is no dataset with id %', new.dataset_id;
+  end if;
+  return new;
+end$$;
+
+comment on function span_integrity_function is e''
+  'Used to simulate a foreign key constraint between span and dataset, to avoid acquiring a lock on the'
+  'dataset table when creating a new partition of span. This function checks that a corresponding dataset'
+  'exists for every inserted or updated span. A trigger that calls this function is added separately to each'
+  'new partition of span.';
+
+create constraint trigger insert_update_span_trigger
+  after insert or update on span
+  for each row
+execute function span_integrity_function();
+
+create or replace procedure span_add_foreign_key_to_partition(table_name varchar)
```

Comment:
Should this just be `create procedure`, and not `create or replace`? Same with `span_integrity_function()` and in general with the added functions in this PR.

Reply:
I suppose the idea is that this code only runs on an empty database - so there shouldn't be a difference (except in cases where we've duplicated a function inadvertently - which would be nice to know about). I can make the change for the new functions - there are likely old functions/triggers around that use `or replace`.

Reply:
Done :+1: